modules/standard/Heap.chpl

```diff
@@ -0,0 +1,400 @@
+          pos = parent;
+        }
+        else break;
+      }
+    }
+
+    pragma "no doc"
+    proc _heapify_down(in pos: int) {
+      while (pos < _data.size) {
+        // find the child node with greater value
+        var greaterChild = pos*2;
+        if (greaterChild >= _data.size) then break; // reach leaf node, break
+        if (greaterChild + 1 < _data.size) {
+          // if the right child node exists
+          if (_greater(_data[greaterChild+1],_data[greaterChild])) {
+            // if the right child is greater, then update the greaterChild
+            greaterChild += 1;
+          }
+        }
+        // if the greaterChild's value is greater than current node, then swap and continue
+        if (_greater(_data[greaterChild],_data[pos])) {
+          _data[greaterChild] <=> _data[pos];
+          pos = greaterChild;
+        }
+        else break;
+      }
+    }
+
+    /*
+      Push an element into the heap
+
+      :arg element: The element that will be pushed
+      :type element: `eltType`
+    */
+    proc push(in element: eltType)
+    lifetime this < element {
+      _enter();
+      _data.append(element);
+      _heapify_up(_data.size-1);
+      _leave();
+    }
+
+    /*
+      Push elements of a list into the heap
+
+      :arg x: The list of which elements is to push.
+      :type x: `list(eltType)`
+    */
+    proc push(ref x: list(eltType)) {
+      for e in x do
+        push(e);
+    }
+
+    /*
+      Push elements in an array into a heap.
+
+      :arg x: The array of which elements is to push.
+      :type x: `[?d] eltType`
+    */
+    proc push(ref x:[?d] eltType) {
+      for e in x do
+        push(e);
+    }
+
+    /*
+      Pop an element and return it.
+
+      :return: the top element
+      :rtype: eltType
+    */
+    proc pop(): eltType {
+      _enter();
+      if (boundsChecking && isEmpty()) {
+        boundsCheckHalt("Called \"heap.pop\" on an empty heap.");
+      }
+      _data(0) <=> _data(_data.size-1);
+      var ret = _data.pop();
+      _heapify_down(0);
+      _leave();
+      return ret;
+    }
+
+    /*
+      Iterate over the elements of this heap
+
+        .. note::
+          Elements in the heap is not sorted
+
+    */
+    iter these() ref {
+      for e in _data {
+        yield e;
+      }
+    }
+
+    proc const toArray(): [] eltType {
+      if !isCopyableType(eltType) then
+        compilerError("toArray() method is not avaliable on a 'heap'",
+                      " with elements of a type that can't be copied, here: ",
+                      eltType: string);
+      _enter();
+      var l: [0..#size] eltType;
+      for i in 0..#size {
+        l[i] = _data[i];
+      }
+      _leave();
+      return l;
+    }
+
+    /*
+      Returns a new DefaultRectangular array containing the
+      elements contained in this heap. The heap will be empty.
+
+      :return: A new DefaultRectangular array.
+    */
+    proc consume(): [] eltType {
```

Comment:
I may be misremembering this, but I think the discussion about `consume` was about an iterator and not a proc?

Reply:
Well, maybe it's me who misunderstood the discussion.

Reply:
Here, I suggested as an alternative to an iterator. https://github.com/chapel-lang/chapel/issues/15667#issuecomment-636998701. Though admittedly, it is only clear with the context.

Reply:
https://github.com/chapel-lang/chapel/pull/15997/commits/c6081bf72f7bec747d2519eb6642ad9e08a793a3